consensus/istanbul/backend/announce.go

```diff
@@ -1305,11 +1064,7 @@ func (sb *Backend) handleEnodeCertificateMsg(_ consensus.Peer, payload []byte) e
 	}
 	logger = logger.New("msg address", msg.Address)
 
-	var enodeCertificate istanbul.EnodeCertificate
-	if err := rlp.DecodeBytes(msg.Msg, &enodeCertificate); err != nil {
-		logger.Warn("Error in decoding received Istanbul Enode Certificate message content", "err", err, "IstanbulMsg", msg.String())
-		return err
-	}
+	enodeCertificate := msg.EnodeCertificate()
```

Comment:
We need some way to handle the case where the message is valid, but not the type we expect. (i.e. someone sends a prepare message instead of an enode certificate.) In this code as it stands, we would actually be vulnerable to an attack where the peer sends an unexpected message, and a nil pointer panic would occur on line 1070.

Reply:
Thanks for pointing this out!

I hadn't realised that the message code is duplicated between the istanbul.Message and the p2p.Msg. In fact the codes are quite tangled some are used just at the p2p level some are used at just the istanbul level and some are used in both levels. So I think the problem here is the case when the p2p message code should correlate to the istanbul message code but it doesn't. What we can do in this case is decode the istanbul.Message and check that its code matches the p2p code, and chuck the message if it doesn't match. I think that would be a way to solve the issue you have pointed out here. Or alternatively we could just process the message based on its inner code and not worry about it matching the p2p code.

Reply:
Checking the message code during RLP decoding makes the most sense to me. The caller should have an idea of what message they expect to get back, and it should be an error if the payload is that that type.